src/OAuth2.php

```diff
@@ -114,11 +130,20 @@ public function fetchAccessToken($authCode, array $params = [])
             'redirect_uri' => $this->getReturnUrl(),
         ];
 
+        if ($this->enablePkce) {
+            $defaultParams['code_verifier'] = $this->getState('authCodeVerifier');
+        }
+
         $request = $this->createRequest()
             ->setMethod('POST')
             ->setUrl($this->tokenUrl)
             ->setData(array_merge($defaultParams, $params));
 
+         // Azure AD will complain if there is no `Origin` header.
+        if ($this->enablePkce) {
+            $request->addHeaders(['Origin' => Url::to('/')]);
```

Comment:
Using the [OpenIdConnect](https://www.yiiframework.com/extension/yiisoft/yii2-authclient/doc/api/2.2/yii-authclient-openidconnect) client with PKCE enabled, I've got an error from WSO2 Identity Server 5.11.0, complaining that the URI in Origin header is invalid.

<details>
<summary>Error log message from WSO2 IS</summary>

```
ERROR {org.wso2.carbon.identity.cors.valve.CORSValve} - CORS management service error when intercepting an HTTP request. org.wso2.carbon.identity.cors.mgt.core.exception.CORSManagementServiceClientException: Bad URI: Missing scheme, such as http or https
	at org.wso2.carbon.identity.cors.mgt.core.model.Origin.<init>(Origin.java:94)
	at org.wso2.carbon.identity.cors.valve.internal.handler.CORSRequestHandler.handleActualRequest(CORSRequestHandler.java:70)
	at org.wso2.carbon.identity.cors.valve.CORSValve.invoke(CORSValve.java:91)
	at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:110)
	at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:102)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:99)
	at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62)
	at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:145)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
	at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:57)
	at org.wso2.carbon.tomcat.ext.valves.RequestEncodingValve.invoke(RequestEncodingValve.java:49)
	at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:126)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Source code originating the exception: https://github.com/wso2/carbon-identity-framework/blob/v5.18.187/components/cors-mgt/org.wso2.carbon.identity.cors.mgt.core/src/main/java/org/wso2/carbon/identity/cors/mgt/core/model/Origin.java#L94
</details>

According to [RFC 6454 section 7.1](https://www.rfc-editor.org/rfc/rfc6454#section-7.1), the Origin header field should be consisted of `scheme "://" host`, optionally followed by `":" port`. Therefore, `/` is not a valid value. Moreover, the identity server also enforce CORS policy by checking the Origin header field when present.

As a workaround at the moment, I have to create a custom subclass from `OpenIdConnect` to override the `fetchAccessToken` method, and copy codes from both the [OpenIdConnect](https://github.com/yiisoft/yii2-authclient/blob/2.2.12/src/OpenIdConnect.php#L288-L294) and [OAuth2](https://github.com/yiisoft/yii2-authclient/blob/2.2.12/src/OAuth2.php#L117-L154) with the lines adding the Origin header removed.

I'd suggest to make the Origin header optional and customizable, e.g. adding a property `$originHeader` that defaults to a valid URI, and skip adding the header when the value is falsy.

Reply:
I also think resolving to an absolute URL will be good, to make it a valid URL. Azure AD must not be checking the content of the header, only that it exists.

```diff
- $request->addHeaders(['Origin' => Url::to('/')]);
+ $request->addHeaders(['Origin' => Url::to('/', true)]);
```

Reply:
In addition to the Origin header, I have also encountered error response `{"error_description":"The client MUST NOT use more than one authentication method in each","error":"invalid_request"}` from WSO IS when calling `OpenIdConnect::fetchAccessToken`. This is due to invalid authorization header as I didn't supplied the client secret when creating the instance of OpenIdConnect. I think that the client secret is optional in the case of authorization code flow with PKCE and therefore the [`applyClientCredentialsToRequest()`](https://github.com/yiisoft/yii2-authclient/blob/2.2.12/src/OpenIdConnect.php#L367) method should only be invoked when the client secret is supplied. Reason as below.

In the [protocol flow of specification of PKCE](https://datatracker.ietf.org/doc/html/rfc7636#section-1.1), client secret is not a mandatory parameter (it's not even mentioned except in the illustration of auth code interception attack). In the section of [Client Authentication](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2.1) for Token Endpoint in OAuth 2.0 specification, it is mentioned that:

> Confidential clients or other clients issued client credentials MUST authenticate with the authorization server as described in [Section 2.3](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3) when making requests to the token endpoint.

Given that PKCE is originated to mitigate potential authorization code interception attacks on native apps that cannot be considered as confidential clients, I would interpret that the basic authentication with client ID and client secret is optional for authorization code flow with PKCE. There are different interpretations in different implementations though, e.g.:

* [Cloudentity](https://docs.authorization.cloudentity.com/features/oauth/client_auth/client_secrets/#when-to-use-client-secret-authentication):
  > On the other hand, client secrets must not be used for public applications, like desktop or mobile apps, which are not able to store their client secrets securely. Users of such applications have direct access to the code of the app, therefore, it can be decompiled. This may lead to a situation that your client secret is compromised and not safe anymore. Such applications should use authentication set to [None and elevate the use of Proof Key for Code Exchange (PKCE)](https://docs.authorization.cloudentity.com/features/oauth/client_auth/auth_none_with_pkce/).
*  Auth0: [`client_secret` is not a parameter in Authorization Code Flow with PKCE](https://auth0.com/docs/api/authentication#authorization-code-flow-with-pkce45)
* AWS Cognito: Authorization header is optional and the example shows that it is [allowed in Authorization Code Flow with PKCE](https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html#post-token-positive-exchanging-authorization-code-grant-with-pkce-for-tokens)

In the case of using Yii2 and this extension for OIDC or OAuth 2, it is most likely that the Yii2 application runs on a server that can be considered as a confidential client. Yet it doesn't limit one to consider its deployment of an Yii2 application as non-confidential, and in such case a client secret should not be issued, while auth code flow with PKCE can still be used just like the case for native mobile apps.

Reply:
@lacek, @AdeAttwood please either do pull requests or create new issues. Comments will be lost here in an already closed pull request.